substrate-node/pallets/pallet-tfgrid/src/migrations/v14.rs

```diff
@@ -0,0 +1,313 @@
+            Err(e) => {
+                info!(
+                    "failed to parse public ips for farm: {:?}, error: {:?}",
+                    k, e
+                )
+            }
+        }
+
+        let new_farm = Farm {
+            version: 4,
+            id: farm.id,
+            name: farm.name,
+            twin_id: farm.twin_id,
+            pricing_policy_id: farm.pricing_policy_id,
+            certification: farm.certification,
+            public_ips,
+            dedicated_farm: farm.dedicated_farm,
+            farming_policy_limits: farm.farming_policy_limits,
+        };
+
+        migrated_count += 1;
+
+        debug!("Farm: {:?} succesfully migrated", k);
+        Some(new_farm)
+    });
+
+    info!(
+        " <<< Farm storage updated! Migrated {} Farms ✅",
+        migrated_count
+    );
+
+    // Update pallet storage version
+    PalletVersion::<T>::set(types::StorageVersion::V14);
+    info!(" <<< Storage version upgraded");
+
+    // Return the weight consumed by the migration.
+    T::DbWeight::get().reads_writes(migrated_count as Weight + 1, migrated_count as Weight + 1)
+}
+
+fn validate_public_config_ip4<T: Config>(
+    config: PubConfigOf<T>,
+) -> Result<Option<PubConfigOf<T>>, Error<T>>
+where
+    IP4<T>: From<<T as Config>::IP4>,
+    GW4<T>: From<<T as Config>::GW4>,
+    IP6<T>: From<<T as Config>::IP6>,
+    GW6<T>: From<<T as Config>::GW6>,
+{
+    let config_ip4: IP4<T> = config
+        .clone()
+        .ip4
+        .ip
+        .try_into()
+        .map_err(|_| Error::<T>::InvalidPublicIP)?;
+
+    let config_gw4: GW4<T> = config
+        .clone()
+        .ip4
+        .gw
+        .try_into()
+        .map_err(|_| Error::<T>::InvalidPublicIP)?;
+
+    let gw4 = IPv4::parse(&config_gw4.0).map_err(|_| Error::<T>::InvalidPublicIP)?;
+    let ip4 = IPv4Cidr::parse(&config_ip4.0).map_err(|_| Error::<T>::InvalidPublicIP)?;
+
+    if gw4.is_public()
+        && gw4.is_unicast()
+        && ip4.is_public()
+        && ip4.is_unicast()
+        && ip4.contains(gw4)
+    {
+        if let Some(ipv6) = config.clone().ip6 {
+            let ip6: IP6<T> = ipv6
+                .ip
+                .try_into()
+                .map_err(|_| Error::<T>::InvalidPublicIP)?;
+            let gw6: GW6<T> = ipv6
+                .gw
+                .try_into()
+                .map_err(|_| Error::<T>::InvalidPublicIP)?;
+
+            let ip6_parsed = IPv6Cidr::parse(&ip6.0).map_err(|_| Error::<T>::InvalidIP6)?;
+            let gw6_parsed = IPv6::parse(&gw6.0).map_err(|_| Error::<T>::InvalidIP6)?;
+            if ip6_parsed.is_public() && ip6_parsed.is_unicast() && gw6_parsed.is_public()
+            // && gw6_parsed.is_unicast()
+            // && gw6_parsed.contains(ip6_parsed)
+            {
+                return Ok(Some(config));
+            } else {
+                return Ok(None);
+            }
+        };
+
+        return Ok(Some(config));
+    } else {
+        return Ok(None);
+    }
```

Comment:
Looking at the logic here and I want to be sure those statements are right:
(1) we look at ip6 config only if ip4 config is valid
(2) even if ip4 config is OK, if we find ip6 config NOT OK we return no config at all 

Also, new style suggestion for readability

```rust
    if !gw4.is_public()
        || !gw4.is_unicast()
        || !ip4.is_public()
        || !ip4.is_unicast()
        || !ip4.contains(gw4)
    {
        return Ok(None);
    }

    if let Some(ipv6) = config.clone().ip6 {
        let ip6: IP6<T> = ipv6
            .ip
            .try_into()
            .map_err(|_| Error::<T>::InvalidPublicIP)?;
        let gw6: GW6<T> = ipv6
            .gw
            .try_into()
            .map_err(|_| Error::<T>::InvalidPublicIP)?;

        let ip6_parsed = IPv6Cidr::parse(&ip6.0).map_err(|_| Error::<T>::InvalidIP6)?;
        let gw6_parsed = IPv6::parse(&gw6.0).map_err(|_| Error::<T>::InvalidIP6)?;

        if !ip6_parsed.is_public() || !ip6_parsed.is_unicast() || !gw6_parsed.is_public()
        // && gw6_parsed.is_unicast()
        // && gw6_parsed.contains(ip6_parsed)
        {
            return Ok(None);
        }
    };

    Ok(Some(config))
```

Reply:
I refactored this migration after you commented :D